### PR TITLE
 Support -ubi suffix starting 8.12.0 and 7.17.16

### PR DESCRIPTION
**Pull request by thbkrkr**
This commit makes the operator use the `-ubi` suffix starting from `8.12.0` and `7.17.16` for the maps image, and for any image when the operator is configured with `--ubi-only`.

### CODE REVIEW

File: pkg/controller/common/container/container.go [hunk 1]
```diff
@@ -53,11 +61,15 @@ const (
...
 	AgentImage            Image = "beats/elastic-agent"
-	MapsImage             Image = "elastic-maps-service/elastic-maps-server-ubi8"
+	MapsImage             Image = "elastic-maps-service/elastic-maps-server"
 	LogstashImage         Image = "logstash/logstash"
 )
 
+var ()
+
 // ImageRepository returns the full container image name by concatenating the current container registry and the image path with the given version.
+// A UBI suffix (-ubi8 or -ubi suffix depending on the version) is prepended to the image name for the maps image,
```
Comment (pebrc) [inline pkg/controller/common/container/container.go:71]:
```suggestion
// A UBI suffix (-ubi8 or -ubi suffix depending on the version) is appended to the image name for the maps image,
```

File: pkg/controller/common/container/container_test.go [hunk 2]
```diff
@@ -66,6 +53,92 @@ func TestImageRepository(t *testing.T) {
...
+		},
+		{
+			name:    "Elasticsearch old 7 image in ubi mode uses old -ubi8 suffix",
+			image:   ElasticsearchImage,
+			version: "7.17.15",
+			suffix:  "-ubi",
+			want:    testRegistry + "/elasticsearch/elasticsearch-ubi8:7.17.15",
+		},
+		{
+			name:       "Maps 7 image with custom repository always uses -ubi suffix",
```
Comment (pebrc) [inline pkg/controller/common/container/container_test.go:87]: Do we know that maps is going to adopt the new convention? (If not we should push for it for consistency)

File: pkg/controller/common/container/container.go [hunk 3]
```diff
@@ -66,9 +78,28 @@ func ImageRepository(img Image, version string) string {
...
+	}
+	// use the global container suffix in non-UBI mode
+	if !UBIOnly {
+		suffix += containerSuffix
+	}
+
+	return fmt.Sprintf("%s/%s%s:%s", containerRegistry, image, suffix, version)
+}
+
+// getUBISuffix returns the UBI suffix to use depending on the given version
```
Comment (barkbay) [inline pkg/controller/common/container/container.go:95]:
```suggestion
// getUBISuffix returns the UBI suffix to use depending on the given version.
```

File: pkg/controller/common/container/container_test.go [hunk 4]
```diff
@@ -66,6 +53,92 @@ func TestImageRepository(t *testing.T) {
...
+			version: "8.12.0",
+			suffix:  "-obi1",
+			want:    testRegistry + "/elastic-maps-service/elastic-maps-server-ubi-obi1:8.12.0",
+		},
+		{
+			name:    "Maps old 8 image with custom suffix",
+			image:   MapsImage,
+			version: "8.11.0",
+			suffix:  "-obi1",
+			want:    testRegistry + "/elastic-maps-service/elastic-maps-server-ubi8-obi1:8.11.0",
```
[the remark this answers is not in the document]
Comment (thbkrkr) [reply pkg/controller/common/container/container_test.go:140]: This was already the behavior but no unit test reflected it.

https://github.com/elastic/cloud-on-k8s/blob/485d5114a50257430cd3f8341527e9b798c3925f/pkg/controller/common/container/container.go#L69-L73

```go
# in pseudo-code
if containerSuffix == ""               { img = "elastic-maps-server-ubi8:8.11.0" }
if containerSuffix == "-ubi8"          { img = "elastic-maps-server-ubi8:8.11.0" }
if containerSuffix == "-somethingelse" { img = "elastic-maps-server-ubi8-somethingelse:8.11.0" }
```

I checked out `main`, added this testcase:

```go
{
	name:    "Maps old 8 image with custom suffix",
	image:   MapsImage,
	version: "8.11.0",
	suffix:  "-obi1",
	want:    testRegistry + "/elastic-maps-service/elastic-maps-server-ubi8-obi1:8.11.0",
},
```

The test passes:

```go
--- PASS: TestImageRepository/Maps_old_8_image_with_custom_suffix (0.00s)
```

File: pkg/controller/common/container/container.go [hunk 5]
```diff
@@ -53,22 +61,42 @@ const (
...
 
 	if containerRepository != "" {
 		image = Image(fmt.Sprintf("%s/%s", containerRepository, img.Name()))
 	}
 
-	// don't double append suffix if already contained as e.g. the case for maps
-	if strings.HasSuffix(string(img), containerSuffix) {
-		return fmt.Sprintf("%s/%s:%s", containerRegistry, image, version)
+	suffix := ""
+	UBIOnly := containerSuffix == UBISuffix
```
Comment (barkbay) [inline pkg/controller/common/container/container.go:80]: nit: not a strong opinion but it feels to me that `UBIOnly` is the boolean flag that should be set by `--ubi-only`.

File: pkg/controller/common/container/container.go [hunk 6]
```diff
@@ -15,6 +17,12 @@ var (
 	containerRegistry   = DefaultContainerRegistry
 	containerRepository = ""
 	containerSuffix     = ""
+
+	major7UbiSuffixMinVersion = version.MinFor(7, 17, 16) // min 7.x to use UBISuffix
+	major8UbiSuffixMinVersion = version.MinFor(8, 12, 0)  // min 8.x to use UBISuffix
+
+	UBISuffix    = "-ubi"  // suffix to use when --ubi-only
+	OldUBISuffix = "-ubi8" // old suffix to use when --ubi-only
```
Comment (barkbay) [inline pkg/controller/common/container/container.go:25]:
```suggestion

const (
	UBISuffix    = "-ubi"  // suffix to use when --ubi-only
	OldUBISuffix = "-ubi8" // old suffix to use when --ubi-only
```

**Review by barkbay — APPROVED** [review]
LGTM

File: pkg/controller/common/container/container_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pebrc) [reply pkg/controller/common/container/container_test.go:142]: You are right I misread the old code.

**Review by pebrc — APPROVED** [review]

File: pkg/controller/common/container/container.go (hunk 5 above)
Comment (thbkrkr) [reply pkg/controller/common/container/container.go]: I renamed `UBIOnly` in `useUBISuffix` in dd2ff5a.

File: pkg/controller/common/container/container.go (hunk 6 above)
Comment (thbkrkr) [reply pkg/controller/common/container/container.go]: dd2ff5a

**Pull request merged**